### PR TITLE
 Variables: variables in Markdown links are not interpolated

### LINKED ISSUE #49759: Variables: variables in Markdown links are not expanded in the text panel
**What happened**:
Variable references such as `${__url_time_range}` are not expanded in Markdown links. It used to work with 8.2.7.

The reproducer below turned into:
```
[Example: from=now-6h&to=now](https://example.com/?$%7B__url_time_range%7D)
```

**What you expected to happen**:
It should have been expanded to:
```
[Example: from=now-6h&to=now](https://example.com/?from=now-6h&to=now)
```

**How to reproduce it (as minimally and precisely as possible)**:
To reproduce, modify the text panel example at https://play.grafana.org/d/bPZNLW_Mz/text-panel?orgId=1&editPanel=2 and add:
```
[Example: ${__url_time_range}](https://example.com/?${__url_time_range})
```

**Anything else we need to know?**:
It was originally reproduced with a variable expansion such as `${foo:queryparam}`.

**Environment**:
- Grafana version: 8.5.3 and 8.5.2
- Data source type & version:
- OS Grafana is installed on: Linux
- User OS & Browser: Firefox 100.0.2 on macOS 12.4
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by polibb**
**What this PR does / why we need it**:

Markdown text in Text Panels does not correctly interpolate variables in links because the [`marked` ](https://marked.js.org/) package we use replaces `{` for `%7B` and `}` for `%7D` in URLs before the interpolation. Try it in [Demo](https://marked.js.org/demo/?text=This%20is%20a%20demo%20of%20a%20variable%20in%20a%20text%3A%20%24%7Bvar%7D%0A%0AThis%20is%20a%20demo%20of%20a%20variable%20in%20a%20URL%3A%20%5BExample%3A%20https%3A%2F%2Fexample.com%2F%3F%24%7Btestavariable%7D%5D(https%3A%2F%2Fexample.com%2F%3F%24%7Btestavariable%7D)%0A%0A%3C%0A%0A%26lt%3B%0A%0A%26%0A%0A%26amp%3B&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22extensions%22%3A%20null%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tokenizer%22%3A%20null%2C%0A%20%22walkTokens%22%3A%20null%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=4.0.8).

Changed sequence of actions to:
1. Interpolates all variables in a text panel before any other changes to the text.
2. Parses text to Markdown, if it needs to.
3. Anything else.

- Added e2e test case for variables in URLs.
- Added e2e test case for the `<` <=> `&lt;` [issue discussed here](https://github.com/grafana/support-escalations/issues/2017)
 
**Which issue(s) this PR fixes**:

Fixes #49759

**Special notes for your reviewer**:

[This comment](https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/text/TextPanel.tsx#L48) is removed because it's not relevant anymore. `noSanitize` option always [does the same thing](https://github.com/grafana/grafana/blob/9dd0eff12a23c7b2daefcadaf5314157f7b6410e/packages/grafana-data/src/text/markdown.ts#L50) as the check `config.disableSanitizeHtml ? content : textUtil.sanitizeTextPanelContent(content)` in [TextPanel](https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/text/TextPanel.tsx#L61) does.

### CODE REVIEW

**Review by ashharrison90 — APPROVED** [review]
i think this looks great 👏

File: public/app/plugins/panel/text/TextPanel.tsx [hunk 1]
```diff
@@ -41,24 +41,34 @@ export class TextPanel extends PureComponent<Props, State> {
   }
 
   prepareHTML(html: string): string {
-    return this.interpolateAndSanitizeString(html);
+    const result = this.interpolateString(html);
```
Comment (axelavargas) [inline public/app/plugins/panel/text/TextPanel.tsx:44]: 🤔 Nit: Can we change the name of this function to something more meaningful, maybe `interpolateVariable`?

**Review by axelavargas — COMMENTED** [review]
Fantastic job with this fix and the approach! 🥳  I am thinking this is a good opportunity to add some unit tests to the `TextPanel` component, especially around the whole` config.disableSanitizeHtml` expected behaviour. I am not sure if in the e2e test we are covering that config setup.

File: public/app/plugins/panel/text/TextPanel.tsx (hunk 1 above)
Comment (polibb) [reply public/app/plugins/panel/text/TextPanel.tsx:44]: yeah, `Variables` :)

**Comment by polibb** [comment]
> Fantastic job with this fix and the approach! 🥳 I am thinking this is a good opportunity to add some unit tests to the `TextPanel` component, especially around the whole` config.disableSanitizeHtml` expected behaviour. I am not sure if in the e2e test we are covering that config setup.

Yes, that's more on the side of checking whether the sanitization is always happening, unless specifically disabled. I would lean to the side of it not being connected to the bug of interpolating variables this PR is about. The interpolation of variabels is well covered in the e2e test and it is better because we don't need to mock the `templateSrv` and the `marked` library.
I will look into adding a unit test for `TextPanel` here using the config setting.

**Comment by polibb** [comment]
> > Fantastic job with this fix and the approach! 🥳 I am thinking this is a good opportunity to add some unit tests to the `TextPanel` component, especially around the whole` config.disableSanitizeHtml` expected behaviour. I am not sure if in the e2e test we are covering that config setup.
> 
> Yes, that's more on the side of checking whether the sanitization is always happening, unless specifically disabled. I would lean to the side of it not being connected to the bug of interpolating variables this PR is about. The interpolation of variabels is well covered in the e2e test and it is better because we don't need to mock the `templateSrv` and the `marked` library. I will look into adding a unit test for `TextPanel` here using the config setting.

Axctually I'll create a new issue in order to test TextPanel well in a separate PR :). Thanks for the suggestion!

**Pull request merged**